ocaml/xapi/xapi_pool.ml

```diff
@@ -3605,3 +3656,223 @@ let set_https_only ~__context ~self:_ ~value =
+        ref
+    | _ ->
+        raise Api_errors.(Server_error (multiple_rpm_gpg_keys_exist, [name]))
+  )
+  |> fun gpg_key_ref ->
+  (* Install the key on other member hosts. *)
+  let other_members =
+    let coordinator = Helpers.get_localhost ~__context in
+    Db.Host.get_all ~__context |> List.filter (fun href -> href <> coordinator)
+  in
+  let funs =
+    List.map
+      (fun host () ->
+        try
+          Helpers.call_api_functions ~__context @@ fun rpc session_id ->
+          Client.Host.install_rpmgpgkey ~rpc ~session_id ~self:host ~name
+            ~pubkey ~fingerprint:md.PubKeyMetaData.fingerprint
+        with e ->
+          error "Error from Host.install_rpmgpgkey on host %s: %s"
+            (Ref.string_of host)
+            (ExnHelper.string_of_exn e) ;
+          raise e
+      )
+      other_members
+  in
+  Helpers.run_in_parallel ~funs ~capacity:16 |> ignore ;
+  gpg_key_ref
+
+let uninstall_rpmgpgkey ~__context ~self ~name =
+  (* This function runs on the coordinator host. *)
+  Gpg.assert_name_is_valid ~name ;
+  Db.Pool.get_repositories ~__context ~self
+  |> List.map (fun self -> Db.Repository.get_gpgkey_path ~__context ~self)
+  |> List.exists (fun p -> p = name)
+  |> function
+  | true ->
+      raise Api_errors.(Server_error (rpm_gpg_key_is_in_use, [name]))
+  | false -> (
+      let expr =
+        let open Db_filter_types in
+        Eq (Field "type", Literal "rpm_pubkey")
+      in
+      Db.Gpg_key.get_refs_where ~__context ~expr
+      |> List.filter_map (fun ref ->
+             let name_in_db = Db.Gpg_key.get_name ~__context ~self:ref in
+             if name = name_in_db then Some ref else None
+         )
+      |> function
+      | [] ->
+          raise
+            Api_errors.(
+              Server_error (rpm_gpg_key_record_does_not_exist, [name])
+            )
+      | refs ->
+          let hosts = Db.Host.get_all ~__context in
+          List.iter
+            (fun self ->
+              Db.Gpg_key.set_uninstalled ~__context ~self ~value:true ;
+              let fingerprint = Db.Gpg_key.get_fingerprint ~__context ~self in
+              let funs =
+                List.map
+                  (fun host () ->
+                    try
+                      Helpers.call_api_functions ~__context
+                      @@ fun rpc session_id ->
+                      Client.Host.uninstall_rpmgpgkey ~rpc ~session_id
+                        ~self:host ~name ~fingerprint
+                    with e ->
+                      error "Error from Host.uninstall_rpmgpgkey on host %s: %s"
+                        (Ref.string_of host)
+                        (ExnHelper.string_of_exn e) ;
+                      raise e
+                  )
+                  hosts
+              in
+              Helpers.run_in_parallel ~funs ~capacity:16 |> ignore ;
+              Db.Gpg_key.destroy ~__context ~self
+            )
+            refs
+    )
+
+let sync_rpmgpgkeys ~__context ~self:_ ~hosts =
```

Comment:
I'm unsure about having such a complex function with such a specific usecase, at this point I'm missing a design.

We have complex machinery around certificate ditribution, it's probably much more complex that what's needed here but it's more principled that this function, and I'd like to see what guarantees are needed here exactly to be able to judge the implementation